node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+
+    let BitfieldGossipMessage {
+        relay_parent,
+        signed_availability,
+    } = message;
+    // concurrently pass on the bitfield distribution to all interested peers
+    let interested_peers = tracker
+        .peer_views
+        .iter()
+        .filter(|(_peerid, view)| view.contains(&relay_parent))
+        .map(|(peerid, _)| peerid.clone())
+        .collect::<Vec<PeerId>>();
+    let message = BitfieldGossipMessage {
+        relay_parent: relay_parent,
+        signed_availability,
+    };
+    let bytes = message.encode();
+    ctx.send_message(AllMessages::NetworkBridge(
+        NetworkBridgeMessage::SendMessage(
+            interested_peers,
+            BitfieldDistribution::PROTOCOL_ID,
+            bytes,
+        ),
+    ))
+    .await?;
+    Ok(())
+}
+
+/// Handle an incoming message from a peer
+async fn process_incoming_peer_message<Context>(
+    ctx: Context,
+    tracker: &mut Tracker,
+    peerid: PeerId,
+    message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+    // we don't care about this, not part of our view
+    if !tracker.view.contains(&message.relay_parent) {
+        return modify_reputiation(ctx, peerid, COST_NOT_INTERESTED).await;
+    }
+
+    // Ignore anything the overseer did not tell this subsystem to work on
+    let mut job_data = tracker.per_relay_parent.get_mut(&message.relay_parent);
+    let job_data: &mut _ = if let Some(ref mut  job_data) = job_data {
+        job_data
+    } else {
+        return modify_reputiation(ctx, peerid, COST_NOT_INTERESTED).await;
+    };
+
+    let validator_set = &job_data.validator_set;
+    if validator_set.len() == 0 {
+        return modify_reputiation(ctx, peerid, COST_MISSING_PEER_SESSION_KEY).await;
+    }
+
+    // check all validators that could have signed this message
+    // @todo there must be a better way figuring this out cheaply
+    let signing_context = job_data.signing_context.clone();
+    if let Some(validator) = validator_set.iter().find(|validator| {
+        message
+            .signed_availability
+            .check_signature(&signing_context, validator)
+            .is_ok()
+    }) {
+        let one_per_validator = &mut (job_data.one_per_validator);
+        // only distribute a message of a validator once
+        if one_per_validator.contains(validator) {
+            return Ok(())
+        }
+        one_per_validator.insert(validator.clone());
+    } else  {
+        return modify_reputiation(ctx, peerid, COST_SIGNATURE_INVALID).await;
+    }
+
+    // passed all conditions, distribute!
+    distribute(ctx, tracker, message).await?;
+
+    Ok(())
+}
+
+/// A gossiped or gossipable signed availability bitfield for a particular relay hash
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct BitfieldGossipMessage {
```

Comment:
This seems like a bit more than an implementation detail; it should probably be mentioned in the guide, and possibly moved to `messages.rs`.

Reply:
Other networking subsystems have a `WireMessage` type which serves this rough purpose. However, it is not exposed publicly beyond the crate.